README.md

```diff
@@ -4,222 +4,1008 @@
+
+The first parameter is the raw string that the user has provided us, and I just explained the second parameter.
+
+So for our `HasExtention` validator, all we do is check if the user's file path ends with `this.wantedExtention`, which we set the value of later.
+
+Now, inside `CatCommand` all we've done is attach our `HasExtention` struct as a UDA (and if you're not familiar with D, congrats, you just made your first UDA!). JCLI will automatically detect that `@HasExtention` is a validator because it follows that convention mentioned just above.
+
+Because D is wonderful, it will automatically generate a constructor for us where the first parameter sets the `wantedExtention` member. So `@HasExtention(".json")` will set the extention we want to `".json"`.
+
+And that's literally all there is to it, let's test:
+
+```bash
+# Passing
+$> ./mytool cat ./dub.json
+[contents of dub.json since validation was a success]
+Program exited with status code 0
+
+# Failing
+$> ./mytool cat ./.gitignore
+ERROR: For positional arg 0(filePath): Expected file to have extention of .json
+Program exited with status code -1
+```
+
+The other type of validation is post-arg-binded validation, which performs validation on the final value provided by an arg binder.
+
+Let's make a validator that ensures that the file is under a certain size:
+
+```d
+struct MaxSize
+{
+    ulong maxSize;
+
+    bool onValidate(File file, ref string error)
+    {
+        error = "File is too large.";
+        return file.size() <= this.maxSize;
+    }
+}
+
+@Command("cat", "Displays the contents of a file.")
+struct CatCommand
+{
+    @CommandPositionalArg(0, "filePath", "The path to the file to display.")
+    @HasExtention(".json")
+    @MaxSize(2)
+    File file;
+
+    // omitted...
+}
+```
+
+The convention for post-arg-binded validation is almost exactly the same as pre-arg-binded validation, it also functions in exactly the same way:
+
+```d
+bool onValidate(<TYPE_OF_VALUE_TO_VALIDATE> value, ref string error);
+```
+
+The only difference is that the first parameter isn't a `string`, but instead the type of value that this validator will work with.
+
+Validators can have different overloads of this function if required. You can even make it a template. JCLI is fine with any of that.
+
+We've set the max size to something really small, so we can easily test that it works:
+
+```bash
+$> ./mytool cat ./dub.json
+ERROR: For positional arg 0(filePath): File is too large.
+Program exited with status code -1
+```
+
+Excellent.
+
+**Word of warning**: Due to various reasons I won't get into, JCLI will silently skip over validators that have incorrect interfaces, so if something isn't working it's likely because JCLI has found an issue with it. Specify the version `JCLI_BinderCompilerErrors` inside of your dub.json/dub.sdl in order to try and attempt to debug why.
```

Comment:
While that is definitely a benefit (and quite a large one), I wonder if it's worth it on its own to change over to it, especially since the error checking for `ArgBinder` could be improved.

The main reason ArgBinder will silently ignore invalid validators is because [of a design choice](https://github.com/BradleyChatha/jcli/blob/master/source/jaster/cli/binder.d#L106) where `ArgBinder.bind` will filter out any UDAs it doesn't understand, so the calling code doesn't need to perform filtering.

These are the two checks that `ArgBinder.bind` uses:

```d
 enum canPreValidate(alias V) = isStruct!V && __traits(compiles, { string s; bool a = typeof(V).init.onPreValidate("", s); });
 enum canValidate(alias V)    = isStruct!V && __traits(compiles, { string s; bool a = typeof(V).init.onValidate(T.init, s); });
```

Which hides/mask certain issues, so things like [version(JCLI_BinderCompilerErrors)](https://github.com/BradleyChatha/jcli/blob/master/source/jaster/cli/binder.d#L150) exist.

But looking at it, I think the better behaviour would be to just see if the UDA has a member named `onValidate` and `onPreValidate`, and just try to call it anyway, and let the compiler dish out an error message for us.

So in other words, if the UDA has `onValidate` and/or `onPreValidate`, then recognise it as a validator regardless and just let the compiler generate an error, instead of skipping over it silently.

Reply:
#25